DOC-1790: Schema Registry Metadata Properties (#1690)
redpanda-data/documentation: modules/manage/pages/schema-reg/schema-reg-overview.adoc

Additionally, you cannot have xref:manage:schema-reg/schema-id-validation.adoc#about-schema-id-validation[schema ID validation] with JSON schemas if the xref:manage:schema-reg/schema-id-validation.adoc#set-subject-name-strategy-per-topic[subject name strategy] _is not_ `TopicNameStrategy`.
endif::[]

== Metadata properties

Schema Registry lets you store and retrieve arbitrary key-value metadata properties alongside schemas. Properties such as `owner`, `team`, or `application.version` travel with the schema through its lifecycle. To register a schema with metadata properties, include a `metadata.properties` object in the request body:

[source,json]
----
POST /subjects/my-topic-value/versions
{
  "schema": "{\"type\":\"string\"}",
  "metadata": {
    "properties": {
      "owner": "platform-team",
      "application.version": "2.1.0"
    }
  }
}
----

> Metadata properties are returned on `GET /subjects/{subject}/versions/{version}` and `GET /schemas/ids/{id}` responses. Using rpk, set metadata with the `--metadata-properties` flag (shorthand: `-p`). The flag accepts `key=value` pairs or a JSON string, and you can pass it multiple times to set multiple properties:

Author: ```suggestion
Metadata properties are returned on `GET /subjects/{subject}/versions/{version}` and `GET /schemas/ids/{id}` responses. Using rpk, set metadata with the `--metadata-properties` flag (shorthand: `-p`). The flag accepts `key=value` pairs or a JSON string (for example, `{"key":"value"}`), and you can pass it multiple times to set multiple properties:
```

[source,bash]
----
# key=value pairs — pass the flag multiple times for multiple properties
rpk registry schema create my-subject --schema schema.avsc \
  --metadata-properties owner=platform-team \
  --metadata-properties env=prod

# JSON string — useful when values contain special characters
rpk registry schema create my-subject --schema schema.avsc \
  --metadata-properties '{"owner":"platform-team","application.version":"2.1.0"}'
----

To view metadata on an existing schema, add `--print-metadata` to `rpk registry schema get`. You can also view metadata properties in {ui}.

When you register a new schema version without a `metadata` field, the new version automatically inherits properties from the most recent version of that subject. To clear inherited metadata, send `"metadata": null`. Registering the same schema definition with different metadata properties creates a new schema version.

[NOTE]
====

> Redpanda supports only `metadata.properties` from the Confluent Data Contracts specification. The following features are not supported: `metadata.tags`, `ruleSet`, and the `defaultMetadata`, `overrideMetadata`, `defaultRuleSet`, and `overrideRuleSet` configuration options.

Author: ```suggestion
Redpanda supports only `metadata.properties` from the Confluent Data Contracts specification. The following configuration options are not supported: 
- `metadata.tags` 
- `ruleSet`
- `defaultMetadata`
- `overrideMetadata` 
- `defaultRuleSet`
- `overrideRuleSet`
```

Additionally, you cannot have xref:manage:schema-reg/schema-id-validation.adoc#about-schema-id-validation[schema ID validation] with JSON schemas if the xref:manage:schema-reg/schema-id-validation.adoc#set-subject-name-strategy-per-topic[subject name strategy] _is not_ `TopicNameStrategy`.
endif::[]

== Metadata properties

> Schema Registry lets you store and retrieve arbitrary key-value metadata properties alongside schemas. Properties such as `owner`, `team`, or `application.version` travel with the schema through its lifecycle. Send a `POST` request to `/subjects/{subject}/versions` with a `metadata.properties` object in the request body:

Reviewer: 'Send a 'POST' To do what?    Needs rewording

Reviewer 2: ```suggestion
Schema Registry lets you store and retrieve arbitrary key-value metadata properties alongside schemas. Properties such as `owner`, `team`, or `application.version` travel with the schema through its lifecycle. You can register a new schema with associated metadata properties by sending a `POST` request to `/subjects/{subject}/versions` with a `metadata.properties` object in the request body:
```

[,json]
----
{
  "schema": "{\"type\":\"string\"}",
  "metadata": {
    "properties": {
      "owner": "platform-team",
      "application.version": "2.1.0"
    }
  }
}
----

Metadata properties are returned on `GET /subjects/{subject}/versions/{version}` and `GET /schemas/ids/{id}` responses. To set metadata using rpk, use the `--metadata-properties` flag (shorthand: `-p`). The flag accepts `key=value` pairs or a JSON string (for example, `{"key":"value"}`), and you can pass it multiple times to set multiple properties:

[,bash]
----
# key=value pairs — pass the flag multiple times for multiple properties
rpk registry schema create my-subject --schema schema.avsc \
  --metadata-properties owner=platform-team \
  --metadata-properties env=prod

# JSON string — useful when values contain special characters
rpk registry schema create my-subject --schema schema.avsc \
  --metadata-properties '{"owner":"platform-team","application.version":"2.1.0"}'
----

To view metadata on an existing schema, add `--print-metadata` to `rpk registry schema get`. You can also view metadata properties in {ui}.

> When you register a new schema version without a `metadata` field, the new version automatically inherits properties from the most recent version of that subject. To clear inherited metadata, send `"metadata": null`. Registering the same schema definition with different metadata properties creates a new schema version.

Reviewer: not quite clear.  You clear it when? when inserting the new version or later?  need an example of clearing inherited metadata to clarify

Reviewer 2: `"metadata": null` does not clear the inherited metadata, in fact it is treated the same as not specifying the metadata at all. Instead it is `"metadata": {}` which clears the metadata.

Perhaps something along these lines:
```suggestion
When you register a new schema version without a `metadata` field, the new version automatically inherits properties from the most recent version of that subject. When registering a new schema, if you want to avoid inheriting the previous version's metadata, you can send `"metadata": {}` to register a schema with explicitly no metadata. Registering the same schema definition with different metadata properties creates a new schema version.
```

[NOTE]
====

> Redpanda supports only `metadata.properties` from the https://docs.confluent.io/platform/current/schema-registry/fundamentals/data-contracts.html[Confluent Data Contracts^] specification. The following features are not supported:

Reviewer 2: @mattschumpert @Feediver1 just double checking, are we good to link out to Confluent docs around this?

Author: I think Claude took liberty to do that here. I don't think we should be linking out to Confluent docs.

* `metadata.tags`
* `ruleSet`
* `defaultMetadata` and `overrideMetadata` (configuration options)
* `defaultRuleSet` and `overrideRuleSet` (configuration options)

> ====

Reviewer 2: One more to add here:
```suggestion
* `compatibilityGroup` (configuration option)
====
```

Additionally, you cannot have xref:manage:schema-reg/schema-id-validation.adoc#about-schema-id-validation[schema ID validation] with JSON schemas if the xref:manage:schema-reg/schema-id-validation.adoc#set-subject-name-strategy-per-topic[subject name strategy] _is not_ `TopicNameStrategy`.
endif::[]

== Metadata properties

Schema Registry lets you store and retrieve arbitrary key-value metadata properties alongside schemas. Properties such as `owner`, `team`, or `application.version` travel with the schema through its lifecycle. You can register a new schema with associated metadata properties by sending a `POST` request to `/subjects/{subject}/versions` with a `metadata.properties` object in the request body:

[,json]
----
{
  "schema": "{\"type\":\"string\"}",
  "metadata": {
    "properties": {
      "owner": "platform-team",
      "application.version": "2.1.0"
    }
  }
}
----

> Metadata properties are returned on `GET /subjects/{subject}/versions/{version}` and `GET /schemas/ids/{id}` responses. To set metadata using rpk, use the `--metadata-properties` flag (shorthand: `-p`). The flag accepts `key=value` pairs or a JSON string (for example, `{"key":"value"}`), and you can pass it multiple times to set multiple properties:

Reviewer 3: I would do either 
a) split this paragraph such that the sentence about GET is closer to the "To view metadata" with rpk info below, as the GET command is the REST API equivalent of `rpk registry schema get`. Then, "To set metadata using rpk" (including the example) should go with the POST description above. 
Or b) add new subheadings to indicate that the grouping of concepts is based on how you view and set metadata (i.e. REST vs rpk). I lean towards a) as I think that's more consistent with the rest of our docs

Author: Done.

> Metadata properties are returned on `GET /subjects/{subject}/versions/{version}` and `GET /schemas/ids/{id}` responses. To set metadata using rpk, use the `--metadata-properties` flag (shorthand: `-p`). The flag accepts `key=value` pairs or a JSON string (for example, `{"key":"value"}`), and you can pass it multiple times to set multiple properties:

Reviewer 3: ```suggestion
Metadata properties are returned on `GET /subjects/{subject}/versions/{version}` and `GET /schemas/ids/{id}` responses. To set metadata using rpk, use the xref:reference:rpk/rpk-registry/rpk-registry-schema-create.adoc[`--metadata-properties`] flag (shorthand: `-p`). The flag accepts `key=value` pairs or a JSON string (for example, `{"key":"value"}`), and you can pass it multiple times to set multiple properties:
```

[,bash]
----
# key=value pairs — pass the flag multiple times for multiple properties
rpk registry schema create my-subject --schema schema.avsc \
  --metadata-properties owner=platform-team \
  --metadata-properties env=prod

# JSON string — useful when values contain special characters
rpk registry schema create my-subject --schema schema.avsc \
  --metadata-properties '{"owner":"platform-team","application.version":"2.1.0"}'
----

> To view metadata on an existing schema, add `--print-metadata` to `rpk registry schema get`. You can also view metadata properties in {ui}.

Reviewer 3: Should the [`rpk registry schema get`](https://deploy-preview-1690--redpanda-docs-preview.netlify.app/current/reference/rpk/rpk-registry/rpk-registry-schema-get/) page also be updated with the new flag?

Author: I checked with @r-vasquez on this:
Hi Joyce, this feature is only relevant to schema `create` and `get` and you won't see it at the root level, here are some examples in the original PR description: https://github.com/redpanda-data/redpanda/pull/29431

> When you register a new schema version without a `metadata` field, the new version automatically inherits properties from the most recent version of that subject. When registering a new schema, if you want to avoid inheriting the previous version's metadata, you can send `"metadata": {}` to register a schema with explicitly no metadata. Registering the same schema definition with different metadata properties creates a new schema version.

Reviewer 3: ```suggestion
When you register a new schema version without a `metadata` field, the new version automatically inherits properties from the most recent version of that subject. To avoid inheriting the previous version's metadata, you can send `"metadata": {}` to register a schema with explicitly no metadata. Registering the same schema definition with different metadata properties creates a new schema version.
```

[NOTE]
====

> Redpanda supports only `metadata.properties` from the Confluent Data Contracts specification. The following features are not supported:

Author: ```suggestion
Redpanda supports only `metadata.properties` from the Confluent Data Contracts specification. The following configuration objects are not supported:
```

> Redpanda supports only `metadata.properties` from the https://docs.confluent.io/platform/current/schema-registry/fundamentals/data-contracts.html[Confluent Data Contracts^] specification. The following features are not supported:

Author: ```suggestion
Redpanda supports only `metadata.properties` from the Confluent Data Contracts specification. The following features are not supported:
```

* `metadata.tags`
* `ruleSet`

> * `defaultMetadata` and `overrideMetadata` (configuration options)

Reviewer 3: Just to clarify - are metadata.tags and ruleSet _not_ configuration options? What are they?

Author: On Confluent's side, they are configuration properties: https://docs.confluent.io/platform/current/schema-registry/fundamentals/data-contracts.html#configuration-enhancements

Author: So they are Confluent Data Contract properties, but not supported for our implementation of SR metadata.

[source,bash]
----
# key=value pairs — pass the flag multiple times for multiple properties
rpk registry schema create my-subject --schema schema.avsc \
  --metadata-properties owner=platform-team \
  --metadata-properties env=prod

# JSON string — useful when values contain special characters
rpk registry schema create my-subject --schema schema.avsc \
  --metadata-properties '{"owner":"platform-team","application.version":"2.1.0"}'
----

To view metadata on an existing schema, add `--print-metadata` to `rpk registry schema get`. You can also view metadata properties in {ui}.

When you register a new schema version without a `metadata` field, the new version automatically inherits properties from the most recent version of that subject. To clear inherited metadata, send `"metadata": null`. Registering the same schema definition with different metadata properties creates a new schema version.

[NOTE]
====

> Redpanda supports only `metadata.properties` from the Confluent Data Contracts specification. The following configuration options are not supported: 

Author: ```suggestion
Redpanda supports only `metadata.properties` from the Confluent Data Contracts specification. The following configuration options are not supported: 

```
